File: benchmarks/java/Algorithms.java
/**
 * Ruva Algorithms Library — Java Reference Implementations
 * 
 * Common algorithms for benchmarking and reference:
 *   - Sorting: QuickSort, MergeSort, HeapSort, RadixSort
 *   - Searching: Binary Search, KMP String Matching
 *   - Graph: BFS, DFS, Dijkstra, Topological Sort
 *   - Math: GCD, Fast Exponentiation, Modular Arithmetic
 */

import java.util.*;

public class Algorithms {

    // ═══════════════════════════════════════════════════════════════════════════
    // Sorting Algorithms
    // ═══════════════════════════════════════════════════════════════════════════

    /** QuickSort — O(n log n) average, O(n²) worst case */
    public static void quickSort(int[] arr, int low, int high) {
        if (low < high) {
            int pivot = partition(arr, low, high);
            quickSort(arr, low, pivot - 1);
            quickSort(arr, pivot + 1, high);
        }
    }

    private static int partition(int[] arr, int low, int high) {
        int pivot = arr[high];
        int i = low - 1;
        for (int j = low; j < high; j++) {
            if (arr[j] < pivot) {
                i++;
                swap(arr, i, j);
            }
        }
        swap(arr, i + 1, high);
        return i + 1;
    }

    /** MergeSort — O(n log n) guaranteed */
    public static void mergeSort(int[] arr, int left, int right) {
        if (left < right) {
            int mid = left + (right - left) / 2;
            mergeSort(arr, left, mid);
            mergeSort(arr, mid + 1, right);
            merge(arr, left, mid, right);
        }
    }

    private static void merge(int[] arr, int left, int mid, int right) {
        int n1 = mid - left + 1;
        int n2 = right - mid;
        int[] L = new int[n1];
        int[] R = new int[n2];
        System.arraycopy(arr, left, L, 0, n1);
        System.arraycopy(arr, mid + 1, R, 0, n2);

        int i = 0, j = 0, k = left;
        while (i < n1 && j < n2) {
            arr[k++] = (L[i] <= R[j]) ? L[i++] : R[j++];
        }
        while (i < n1) arr[k++] = L[i++];
        while (j < n2) arr[k++] = R[j++];
    }

    /** HeapSort — O(n log n) in-place */
    public static void heapSort(int[] arr) {
        int n = arr.length;
        for (int i = n / 2 - 1; i >= 0; i--) heapify(arr, n, i);
        for (int i = n - 1; i > 0; i--) {
            swap(arr, 0, i);
            heapify(arr, i, 0);
        }
    }

    private static void heapify(int[] arr, int n, int i) {
        int largest = i, left = 2 * i + 1, right = 2 * i + 2;
        if (left < n && arr[left] > arr[largest]) largest = left;
        if (right < n && arr[right] > arr[largest]) largest = right;
        if (largest != i) {
            swap(arr, i, largest);
            heapify(arr, n, largest);
        }
    }

    /** RadixSort — O(d × (n + b)) where d = digits, b = base */
    public static void radixSort(int[] arr) {
        int max = Arrays.stream(arr).max().orElse(0);
        for (int exp = 1; max / exp > 0; exp *= 10) {
            countingSortByDigit(arr, exp);
        }
    }

    private static void countingSortByDigit(int[] arr, int exp) {
        int n = arr.length;
        int[] output = new int[n];
        int[] count = new int[10];

        for (int val : arr) count[(val / exp) % 10]++;
        for (int i = 1; i < 10; i++) count[i] += count[i - 1];
        for (int i = n - 1; i >= 0; i--) {
            output[--count[(arr[i] / exp) % 10]] = arr[i];
        }
        System.arraycopy(output, 0, arr, 0, n);
    }

    // ═══════════════════════════════════════════════════════════════════════════
    // Searching Algorithms
    // ═══════════════════════════════════════════════════════════════════════════

    /** Binary Search — O(log n) on sorted array */
    public static int binarySearch(int[] arr, int target) {
        int lo = 0, hi = arr.length - 1;
        while (lo <= hi) {
            int mid = lo + (hi - lo) / 2;
            if (arr[mid] == target) return mid;
            else if (arr[mid] < target) lo = mid + 1;
            else hi = mid - 1;
        }
        return -1;
    }

    /** KMP String Search — O(n + m) */
    public static int kmpSearch(String text, String pattern) {
        int[] lps = computeLPS(pattern);
        int i = 0, j = 0;
        while (i < text.length()) {
            if (text.charAt(i) == pattern.charAt(j)) {
                i++;
                j++;
                if (j == pattern.length()) return i - j;
            } else if (j > 0) {
                j = lps[j - 1];
            } else {
                i++;
            }
        }
        return -1;
    }

    private static int[] computeLPS(String pattern) {
        int[] lps = new int[pattern.length()];
        int len = 0, i = 1;
        while (i < pattern.length()) {
            if (pattern.charAt(i) == pattern.charAt(len)) {
                lps[i++] = ++len;
            } else if (len > 0) {
                len = lps[len - 1];
            } else {
                lps[i++] = 0;
            }
        }
        return lps;
    }

    // ═══════════════════════════════════════════════════════════════════════════
    // Graph Algorithms
    // ═══════════════════════════════════════════════════════════════════════════

    /** BFS — Breadth-First Search */
    public static List<Integer> bfs(List<List<Integer>> adj, int start) {
        List<Integer> result = new ArrayList<>();
        boolean[] visited = new boolean[adj.size()];
        Queue<Integer> queue = new LinkedList<>();
        visited[start] = true;
        queue.add(start);

        while (!queue.isEmpty()) {
            int node = queue.poll();
            result.add(node);
            for (int neighbor : adj.get(node)) {
                if (!visited[neighbor]) {
                    visited[neighbor] = true;
                    queue.add(neighbor);
                }
            }
        }
        return result;
    }

    /** DFS — Depth-First Search (recursive) */
    public static List<Integer> dfs(List<List<Integer>> adj, int start) {
        List<Integer> result = new ArrayList<>();
        boolean[] visited = new boolean[adj.size()];
        dfsHelper(adj, start, visited, result);
        return result;
    }

    private static void dfsHelper(List<List<Integer>> adj, int node,
                                   boolean[] visited, List<Integer> result) {
        visited[node] = true;
        result.add(node);
        for (int neighbor : adj.get(node)) {
            if (!visited[neighbor]) {
                dfsHelper(adj, neighbor, visited, result);
            }
        }
    }

    /** Dijkstra's Shortest Path — O((V + E) log V) */
    public static int[] dijkstra(List<List<int[]>> adj, int src) {
        int n = adj.size();
        int[] dist = new int[n];
        Arrays.fill(dist, Integer.MAX_VALUE);
        dist[src] = 0;

        PriorityQueue<int[]> pq = new PriorityQueue<>(Comparator.comparingInt(a -> a[1]));
        pq.add(new int[]{src, 0});

        while (!pq.isEmpty()) {
            int[] curr = pq.poll();
            int u = curr[0], d = curr[1];
            if (d > dist[u]) continue;
            for (int[] edge : adj.get(u)) {
                int v = edge[0], w = edge[1];
                if (dist[u] + w < dist[v]) {
                    dist[v] = dist[u] + w;
                    pq.add(new int[]{v, dist[v]});
                }
            }
        }
        return dist;
    }

    /** Topological Sort — Kahn's algorithm */
    public static List<Integer> topologicalSort(List<List<Integer>> adj) {
        int n = adj.size();
        int[] inDegree = new int[n];
        for (int u = 0; u < n; u++) {
            for (int v : adj.get(u)) inDegree[v]++;
        }

        Queue<Integer> queue = new LinkedList<>();
        for (int i = 0; i < n; i++) {
            if (inDegree[i] == 0) queue.add(i);
        }

        List<Integer> result = new ArrayList<>();
        while (!queue.isEmpty()) {
            int u = queue.poll();
            result.add(u);
            for (int v : adj.get(u)) {
                if (--inDegree[v] == 0) queue.add(v);
            }
        }

        if (result.size() != n) {
            throw new IllegalStateException("Graph has a cycle");
        }
        return result;
    }

    // ═══════════════════════════════════════════════════════════════════════════
    // Math Utilities
    // ═══════════════════════════════════════════════════════════════════════════

    /** Greatest Common Divisor */
    public static long gcd(long a, long b) {
        while (b != 0) {
            long temp = b;
            b = a % b;
            a = temp;
        }
        return a;
    }

    /** Fast modular exponentiation — O(log n) */
    public static long modPow(long base, long exp, long mod) {
        long result = 1;
        base %= mod;
        while (exp > 0) {
            if ((exp & 1) == 1) {
                result = (result * base) % mod;
            }
            exp >>= 1;
            base = (base * base) % mod;
        }
        return result;
    }

    /** Extended Euclidean Algorithm — returns {gcd, x, y} such that ax + by = gcd */
    public static long[] extendedGcd(long a, long b) {
        if (b == 0) return new long[]{a, 1, 0};
        long[] prev = extendedGcd(b, a % b);
        long gcd = prev[0];
        long x = prev[2];
        long y = prev[1] - (a / b) * prev[2];
        return new long[]{gcd, x, y};
    }

    /** Sieve of Eratosthenes — returns list of primes up to limit */
    public static List<Integer> sieveOfEratosthenes(int limit) {
        boolean[] sieve = new boolean[limit + 1];
        Arrays.fill(sieve, true);
        sieve[0] = sieve[1] = false;

        for (int p = 2; p * p <= limit; p++) {
            if (sieve[p]) {
                for (int i = p * p; i <= limit; i += p) {
                    sieve[i] = false;
                }
            }
        }

        List<Integer> primes = new ArrayList<>();
        for (int i = 2; i <= limit; i++) {
            if (sieve[i]) primes.add(i);
        }
        return primes;
    }

    // ═══════════════════════════════════════════════════════════════════════════
    // Helpers
    // ═══════════════════════════════════════════════════════════════════════════

    private static void swap(int[] arr, int i, int j) {
        int tmp = arr[i];
        arr[i] = arr[j];
        arr[j] = tmp;
    }

    // ═══════════════════════════════════════════════════════════════════════════
    // Demo / Test
    // ═══════════════════════════════════════════════════════════════════════════

    public static void main(String[] args) {
        System.out.println("=== Ruva Algorithms Library (Java) ===\n");

        // Sorting tests
        int[] data = {38, 27, 43, 3, 9, 82, 10};
        int[] copy;

        copy = data.clone();
        quickSort(copy, 0, copy.length - 1);
        System.out.println("QuickSort:    " + Arrays.toString(copy));

        copy = data.clone();
        mergeSort(copy, 0, copy.length - 1);
        System.out.println("MergeSort:    " + Arrays.toString(copy));

        copy = data.clone();
        heapSort(copy);
        System.out.println("HeapSort:     " + Arrays.toString(copy));

        copy = data.clone();
        radixSort(copy);
        System.out.println("RadixSort:    " + Arrays.toString(copy));

        // Searching
        int[] sorted = {1, 3, 5, 7, 9, 11, 13};
        System.out.println("\nBinarySearch(7): index=" + binarySearch(sorted, 7));
        System.out.println("KMP('ABABD', 'ABD'): index=" + kmpSearch("ABABDABACDABABCABAB", "ABD"));

        // Graph
        List<List<Integer>> graph = new ArrayList<>();
        for (int i = 0; i < 5; i++) graph.add(new ArrayList<>());
        graph.get(0).addAll(Arrays.asList(1, 2));
        graph.get(1).addAll(Arrays.asList(3, 4));
        graph.get(2).add(4);

        System.out.println("\nBFS from 0:  " + bfs(graph, 0));
        System.out.println("DFS from 0:  " + dfs(graph, 0));

        // Math
        System.out.println("\ngcd(48, 18):  " + gcd(48, 18));
        System.out.println("modPow(2,10,1000): " + modPow(2, 10, 1000));
        System.out.println("Primes <= 50: " + sieveOfEratosthenes(50).size() + " primes");

        System.out.println("\nAll algorithms functional.");
    }
}

File: benchmarks/python/algorithms.py
#!/usr/bin/env python3
"""
Ruva Algorithms Library — Python Reference Implementations

Common algorithms for benchmarking and reference:
  - Sorting: QuickSort, MergeSort, HeapSort, RadixSort, TimSort
  - Searching: Binary Search, KMP String Matching
  - Graph: BFS, DFS, Dijkstra, Topological Sort
  - Math: GCD, Fast Exponentiation, Modular Arithmetic, Sieve
  - Dynamic Programming: Fibonacci, LCS, Knapsack
"""

from __future__ import annotations
import time
import math
from collections import deque
from typing import Any, Callable, List, Optional, Tuple


# ═══════════════════════════════════════════════════════════════════════════
# Sorting Algorithms
# ═══════════════════════════════════════════════════════════════════════════

def quick_sort(arr: list[int]) -> list[int]:
    """QuickSort — O(n log n) average."""
    if len(arr) <= 1:
        return arr
    pivot = arr[len(arr) // 2]
    left = [x for x in arr if x < pivot]
    middle = [x for x in arr if x == pivot]
    right = [x for x in arr if x > pivot]
    return quick_sort(left) + middle + quick_sort(right)


def merge_sort(arr: list[int]) -> list[int]:
    """MergeSort — O(n log n) guaranteed."""
    if len(arr) <= 1:
        return arr
    mid = len(arr) // 2
    left = merge_sort(arr[:mid])
    right = merge_sort(arr[mid:])
    return _merge(left, right)


def _merge(left: list[int], right: list[int]) -> list[int]:
    result: list[int] = []
    i = j = 0
    while i < len(left) and j < len(right):
        if left[i] <= right[j]:
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1
    result.extend(left[i:])
    result.extend(right[j:])
    return result


def heap_sort(arr: list[int]) -> list[int]:
    """HeapSort — O(n log n) using heapq."""
    import heapq
    heap = list(arr)
    heapq.heapify(heap)
    return [heapq.heappop(heap) for _ in range(len(heap))]


def radix_sort(arr: list[int]) -> list[int]:
    """RadixSort — O(d × (n + b)) using counting sort by digit."""
    if not arr:
        return arr

    max_val = max(abs(x) for x in arr)
    exp = 1
    result = list(arr)

    while max_val // exp > 0:
        result = _counting_sort_by_digit(result, exp)
        exp *= 10

    return result


def _counting_sort_by_digit(arr: list[int], exp: int) -> list[int]:
    n = len(arr)
    output = [0] * n
    count = [0] * 10

    for val in arr:
        digit = abs(val) // exp % 10
        count[digit] += 1

    for i in range(1, 10):
        count[i] += count[i - 1]

    for i in range(n - 1, -1, -1):
        digit = abs(arr[i]) // exp % 10
        count[digit] -= 1
        output[count[digit]] = arr[i]

    return output


def tim_sort(arr: list[int]) -> list[int]:
    """TimSort — Python's built-in Timsort wrapper."""
    return sorted(arr)


# ═══════════════════════════════════════════════════════════════════════════
# Searching Algorithms
# ═══════════════════════════════════════════════════════════════════════════

def binary_search(arr: list[int], target: int) -> int:
    """Binary Search — O(log n) on sorted array. Returns index or -1."""
    lo, hi = 0, len(arr) - 1
    while lo <= hi:
        mid = lo + (hi - lo) // 2
        if arr[mid] == target:
            return mid
        elif arr[mid] < target:
            lo = mid + 1
        else:
            hi = mid - 1
    return -1


def kmp_search(text: str, pattern: str) -> int:
    """KMP String Search — O(n + m). Returns index or -1."""
    if not pattern:
        return 0
    lps = _compute_lps(pattern)
    i = j = 0

    while i < len(text):
        if text[i] == pattern[j]:
            i += 1
            j += 1
            if j == len(pattern):
                return i - j
        elif j > 0:
            j = lps[j - 1]
        else:
            i += 1

    return -1


def _compute_lps(pattern: str) -> list[int]:
    lps = [0] * len(pattern)
    length = 0
    i = 1
    while i < len(pattern):
        if pattern[i] == pattern[length]:
            length += 1
            lps[i] = length
            i += 1
        elif length > 0:
            length = lps[length - 1]
        else:
            lps[i] = 0
            i += 1
    return lps


# ═══════════════════════════════════════════════════════════════════════════
# Graph Algorithms
# ═══════════════════════════════════════════════════════════════════════════

def bfs(adj: list[list[int]], start: int) -> list[int]:
    """BFS — Breadth-First Search."""
    visited = [False] * len(adj)
    queue: deque[int] = deque([start])
    visited[start] = True
    result: list[int] = []

    while queue:
        node = queue.popleft()
        result.append(node)
        for neighbor in adj[node]:
            if not visited[neighbor]:
                visited[neighbor] = True
                queue.append(neighbor)

    return result


def dfs(adj: list[list[int]], start: int) -> list[int]:
    """DFS — Depth-First Search (iterative)."""
    visited = [False] * len(adj)
    stack = [start]
    result: list[int] = []

    while stack:
        node = stack.pop()
        if visited[node]:
            continue
        visited[node] = True
        result.append(node)
        for neighbor in reversed(adj[node]):
            if not visited[neighbor]:
                stack.append(neighbor)

    return result


def dijkstra(adj: list[list[Tuple[int, int]]], src: int) -> list[int]:
    """Dijkstra's Shortest Path — O((V + E) log V)."""
    import heapq
    n = len(adj)
    dist = [float("inf")] * n
    dist[src] = 0
    pq: list[Tuple[int, int]] = [(0, src)]

    while pq:
        d, u = heapq.heappop(pq)
        if d > dist[u]:
            continue
        for v, w in adj[u]:
            if dist[u] + w < dist[v]:
                dist[v] = dist[u] + w
                heapq.heappush(pq, (dist[v], v))

    return dist


def topological_sort(adj: list[list[int]]) -> list[int]:
    """Topological Sort — Kahn's algorithm."""
    n = len(adj)
    in_degree = [0] * n
    for u in range(n):
        for v in adj[u]:
            in_degree[v] += 1

    queue: deque[int] = deque()
    for i in range(n):
        if in_degree[i] == 0:
            queue.append(i)

    result: list[int] = []
    while queue:
        u = queue.popleft()
        result.append(u)
        for v in adj[u]:
            in_degree[v] -= 1
            if in_degree[v] == 0:
                queue.append(v)

    if len(result) != n:
        raise ValueError("Graph has a cycle")
    return result


# ═══════════════════════════════════════════════════════════════════════════
# Math Utilities
# ═══════════════════════════════════════════════════════════════════════════

def gcd(a: int, b: int) -> int:
    """Greatest Common Divisor."""
    while b:
        a, b = b, a % b
    return a


def mod_pow(base: int, exp: int, mod: int) -> int:
    """Fast modular exponentiation — O(log n)."""
    result = 1
    base %= mod
    while exp > 0:
        if exp & 1:
            result = result * base % mod
        exp >>= 1
        base = base * base % mod
    return result


def extended_gcd(a: int, b: int) -> Tuple[int, int, int]:
    """Extended Euclidean Algorithm — returns (gcd, x, y) such that ax + by = gcd."""
    if b == 0:
        return a, 1, 0
    gcd_val, x, y = extended_gcd(b, a % b)
    return gcd_val, y, x - (a // b) * y


def sieve_of_eratosthenes(limit: int) -> list[int]:
    """Sieve of Eratosthenes — returns list of primes up to limit."""
    sieve = [True] * (limit + 1)
    sieve[0] = sieve[1] = False

    p = 2
    while p * p <= limit:
        if sieve[p]:
            for i in range(p * p, limit + 1, p):
                sieve[i] = False
        p += 1

    return [i for i, is_prime in enumerate(sieve) if is_prime]


# ═══════════════════════════════════════════════════════════════════════════
# Dynamic Programming
# ═══════════════════════════════════════════════════════════════════════════

def fibonacci_dp(n: int) -> int:
    """Fibonacci with bottom-up DP — O(n) time, O(1) space."""
    if n <= 1:
        return n
    a, b = 0, 1
    for _ in range(2, n + 1):
        a, b = b, a + b
    return b


def longest_common_subsequence(s1: str, s2: str) -> int:
    """LCS — O(mn) time, O(mn) space."""
    m, n = len(s1), len(s2)
    dp = [[0] * (n + 1) for _ in range(m + 1)]

    for i in range(1, m + 1):
        for j in range(1, n + 1):
            if s1[i - 1] == s2[j - 1]:
                dp[i][j] = dp[i - 1][j - 1] + 1
            else:
                dp[i][j] = max(dp[i - 1][j], dp[i][j - 1])

    return dp[m][n]


def knapsack_01(weights: list[int], values: list[int], capacity: int) -> int:
    """0/1 Knapsack — O(n × W) time."""
    n = len(weights)
    dp = [0] * (capacity + 1)

    for i in range(n):
        for w in range(capacity, weights[i] - 1, -1):
            dp[w] = max(dp[w], dp[w - weights[i]] + values[i])

    return dp[capacity]


# ═══════════════════════════════════════════════════════════════════════════
# Demo / Test
# ═══════════════════════════════════════════════════════════════════════════

def main() -> None:
    print("=== Ruva Algorithms Library (Python) ===\n")

    data = [38, 27, 43, 3, 9, 82, 10]

    # Sorting
    print(f"QuickSort:    {quick_sort(data)}")
    print(f"MergeSort:    {merge_sort(data)}")
    print(f"HeapSort:     {heap_sort(data)}")
    print(f"RadixSort:    {radix_sort(data)}")
    print(f"TimSort:      {tim_sort(data)}")

    # Searching
    sorted_arr = [1, 3, 5, 7, 9, 11, 13]
    print(f"\nBinarySearch(7): index={binary_search(sorted_arr, 7)}")
    print(f"KMP('ABABD', 'ABD'): index={kmp_search('ABABDABACDABABCABAB', 'ABD')}")

    # Graph
    graph: list[list[int]] = [[] for _ in range(5)]
    graph[0].extend([1, 2])
    graph[1].extend([3, 4])
    graph[2].append(4)

    print(f"\nBFS from 0:  {bfs(graph, 0)}")
    print(f"DFS from 0:  {dfs(graph, 0)}")

    # Math
    print(f"\ngcd(48, 18):  {gcd(48, 18)}")
    print(f"modPow(2,10,1000): {mod_pow(2, 10, 1000)}")
    print(f"Primes <= 50: {len(sieve_of_eratosthenes(50))} primes")

    # DP
    print(f"\nfibonacci(40): {fibonacci_dp(40)}")
    print(f"LCS('ABCBDAB', 'BDCAB'): {longest_common_subsequence('ABCBDAB', 'BDCAB')}")
    print(f"Knapsack(4 items, W=7): {knapsack_01([1, 3, 4, 5], [1, 4, 5, 7], 7)}")

    print("\nAll algorithms functional.")


if __name__ == "__main__":
    main()
